File: legacy-python/utils/agent_manager.py
"""Agent management utilities for the DAO simulation."""

import random
from typing import Dict, List, Type, Any, Optional

from agents import (
    AdaptiveInvestor,
    Arbitrator,
    Artist,
    Auditor,
    BountyHunter,
    Collector,
    Delegator,
    Developer,
    ExternalPartner,
    Investor,
    LiquidDelegator,
    PassiveMember,
    ProposalCreator,
    RLTrader,
    Regulator,
    ServiceProvider,
    Speculator,
    Trader,
    Validator,
)
from constants import (
    ADAPTIVE_INVESTOR_BUDGET,
    ADAPTIVE_INVESTOR_TOKENS,
    ARBITRATOR_CAPACITY,
    ARBITRATOR_TOKENS,
    ARTIST_TOKENS,
    AUDITOR_TOKENS,
    BOUNTY_HUNTER_REPUTATION,
    BOUNTY_HUNTER_TOKENS,
    COLLECTOR_TOKENS,
    DEFAULT_AGENT_REPUTATION,
    DEFAULT_AGENT_TOKENS,
    DELEGATOR_BUDGET,
    DELEGATOR_TOKENS,
    DEVELOPER_TOKENS,
    EXTERNAL_PARTNER_TOKENS,
    INVESTOR_BUDGET,
    INVESTOR_TOKENS,
    LIQUID_DELEGATOR_BUDGET,
    LIQUID_DELEGATOR_TOKENS,
    NEW_MEMBER_COST,
    NEW_MEMBER_FUND_REQUIREMENT,
    NEW_MEMBER_INTERVAL,
    NEW_MEMBERS_PER_INTERVAL,
    PASSIVE_MEMBER_TOKENS,
    PROPOSAL_CREATOR_TOKENS,
    REGULATOR_TOKENS,
    SERVICE_PROVIDER_BUDGET,
    SERVICE_PROVIDER_TOKENS,
    SPECULATOR_TOKENS,
    TRADER_TOKENS,
    VALIDATOR_TOKENS,
)
from utils.locations import generate_random_location


class AgentManager:
    """Manages agent creation, caching, and lifecycle operations."""
    
    def __init__(self, dao_simulation):
        self.simulation = dao_simulation
        self._agent_type_cache: Dict[str, List[Any]] = {}
        self._cache_dirty = True
        
        # Agent configuration templates
        self.agent_configs = {
            Developer: {
                "tokens": DEVELOPER_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION,
                "skillset": ["Python", "JavaScript"]
            },
            Investor: {
                "tokens": INVESTOR_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION,
                "investment_budget": INVESTOR_BUDGET
            },
            Trader: {
                "tokens": TRADER_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION
            },
            AdaptiveInvestor: {
                "tokens": ADAPTIVE_INVESTOR_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION,
                "investment_budget": ADAPTIVE_INVESTOR_BUDGET,
            },
            Delegator: {
                "tokens": DELEGATOR_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION,
                "delegation_budget": DELEGATOR_BUDGET,
            },
            LiquidDelegator: {
                "tokens": LIQUID_DELEGATOR_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION,
                "delegation_budget": LIQUID_DELEGATOR_BUDGET,
            },
            ProposalCreator: {
                "tokens": PROPOSAL_CREATOR_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION
            },
            Validator: {
                "tokens": VALIDATOR_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION
            },
            ServiceProvider: {
                "tokens": SERVICE_PROVIDER_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION,
                "service_budget": SERVICE_PROVIDER_BUDGET,
            },
            Arbitrator: {
                "tokens": ARBITRATOR_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION,
                "arbitration_capacity": ARBITRATOR_CAPACITY,
            },
            Regulator: {
                "tokens": REGULATOR_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION
            },
            Auditor: {
                "tokens": AUDITOR_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION
            },
            BountyHunter: {
                "tokens": BOUNTY_HUNTER_TOKENS,
                "reputation": BOUNTY_HUNTER_REPUTATION
            },
            ExternalPartner: {
                "tokens": EXTERNAL_PARTNER_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION,
                "voting_strategy": None,
            },
            Artist: {
                "tokens": ARTIST_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION
            },
            Collector: {
                "tokens": COLLECTOR_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION
            },
            Speculator: {
                "tokens": SPECULATOR_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION
            },
            PassiveMember: {
                "tokens": PASSIVE_MEMBER_TOKENS,
                "reputation": DEFAULT_AGENT_REPUTATION
            },
        }
    
    def get_agents_by_type(self, agent_type: str) -> List[Any]:
        """Get agents of a specific type using cache for performance."""
        if self._cache_dirty or agent_type not in self._agent_type_cache:
            self._rebuild_agent_cache()
        return self._agent_type_cache.get(agent_type, [])
    
    def _rebuild_agent_cache(self) -> None:
        """Rebuild the agent type cache."""
        self._agent_type_cache.clear()
        for agent in self.simulation.dao.members:
            agent_class = type(agent).__name__
            if agent_class not in self._agent_type_cache:
                self._agent_type_cache[agent_class] = []
            self._agent_type_cache[agent_class].append(agent)
        self._cache_dirty = False
    
    def invalidate_cache(self) -> None:
        """Mark cache as dirty - should be called when agents are added/removed."""
        self._cache_dirty = True
    
    def create_agent(self, agent_class: Type, agent_id: str, **kwargs) -> Any:
        """Create a new agent with appropriate configuration."""
        base_config = self.agent_configs.get(agent_class, {
            "tokens": DEFAULT_AGENT_TOKENS,
            "reputation": DEFAULT_AGENT_REPUTATION
        })
        
        # Merge base config with provided kwargs
        config = {**base_config, **kwargs}
        
        return agent_class(
            unique_id=agent_id,
            model=self.simulation,
            location=generate_random_location(),
            **config
        )
    
    def add_new_members(self) -> None:
        """Add new members periodically if treasury has sufficient funds."""
        agent_classes = [
            Developer, Investor, Trader, AdaptiveInvestor, Delegator,
            LiquidDelegator, ProposalCreator, Validator, ServiceProvider,
            Arbitrator, Regulator, Auditor, BountyHunter, ExternalPartner,
            Artist, Collector, Speculator, PassiveMember
        ]
        
        for _ in range(NEW_MEMBERS_PER_INTERVAL):
            if (self.simulation.schedule.steps > 0 and 
                self.simulation.schedule.steps % NEW_MEMBER_INTERVAL == 0):
                if self.simulation.dao.treasury.funds >= NEW_MEMBER_FUND_REQUIREMENT:
                    agent_class = random.choice(agent_classes)
                    new_agent = self.create_agent(
                        agent_class, 
                        f"{agent_class.__name__}_{self.simulation.schedule.steps}"
                    )
                    
                    if new_agent.reputation > 25:  # Add only if reputation is above 25
                        self.simulation.dao.add_member(new_agent)
                        self.simulation.schedule.add(new_agent)
                        self.invalidate_cache()
                        
                        # Add new agent to the space for visualization
                        if self.simulation.space is not None:
                            import random
                            x = random.randint(0, self.simulation.space.width - 1)
                            y = random.randint(0, self.simulation.space.height - 1)
                            self.simulation.space.place_agent(new_agent, (x, y))
                        
                        self.simulation.dao.treasury.withdraw(
                            "DAO_TOKEN", NEW_MEMBER_COST, step=self.simulation.schedule.steps
                        )
    
    def cull_members(self) -> None:
        """Remove members with negative reputation."""
        agents_to_remove = []
        for agent in self.simulation.dao.members:
            if agent.reputation < 0:
                agents_to_remove.append(agent)
        
        for agent in agents_to_remove:
            self.simulation.dao.remove_member(agent)
            self.simulation.schedule.remove(agent)
            self.invalidate_cache()
            
            # Remove agent from the space for visualization
            if self.simulation.space is not None:
                self.simulation.space.remove_agent(agent)

File: lib/utils/agent-manager.ts
// Agent Manager - manages agent creation, caching, and lifecycle
// Port from utils/agent_manager.py

import type { DAOModel } from '../engine/model';
import type { DAOMember } from '../agents/base';
import {
  Developer,
  Investor,
  Trader,
  AdaptiveInvestor,
  Delegator,
  LiquidDelegator,
  ProposalCreator,
  Validator,
  ServiceProvider,
  Arbitrator,
  Regulator,
  Auditor,
  BountyHunter,
  ExternalPartner,
  Artist,
  Collector,
  Speculator,
  PassiveMember,
  StakerAgent,
  GovernanceWhale,
} from '../agents';
import { generateRandomLocation } from './locations';
import * as constants from '../config/constants';
import { randomChoice } from './random';

type AgentClass = new (
  uniqueId: string,
  model: DAOModel,
  ...args: any[]
) => DAOMember;

interface AgentConfig {
  tokens: number;
  reputation: number;
  [key: string]: any;
}

export class AgentManager {
  simulation: DAOModel;
  private agentTypeCache: Map<string, DAOMember[]> = new Map();
  private cacheDirty: boolean = true;

  // Agent configuration templates
  agentConfigs: Map<AgentClass, AgentConfig> = new Map<AgentClass, AgentConfig>([
    [Developer as AgentClass, {
      tokens: constants.DEVELOPER_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
      skillset: ['TypeScript', 'JavaScript', 'Solidity'],
    }],
    [Investor as AgentClass, {
      tokens: constants.INVESTOR_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
      investmentBudget: constants.INVESTOR_BUDGET,
    }],
    [Trader as AgentClass, {
      tokens: constants.TRADER_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
    }],
    [AdaptiveInvestor as AgentClass, {
      tokens: constants.ADAPTIVE_INVESTOR_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
      investmentBudget: constants.ADAPTIVE_INVESTOR_BUDGET,
    }],
    [Delegator as AgentClass, {
      tokens: constants.DELEGATOR_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
      delegationBudget: constants.DELEGATOR_BUDGET,
    }],
    [LiquidDelegator as AgentClass, {
      tokens: constants.LIQUID_DELEGATOR_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
      delegationBudget: constants.LIQUID_DELEGATOR_BUDGET,
    }],
    [ProposalCreator as AgentClass, {
      tokens: constants.PROPOSAL_CREATOR_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
    }],
    [Validator as AgentClass, {
      tokens: constants.VALIDATOR_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
    }],
    [ServiceProvider as AgentClass, {
      tokens: constants.SERVICE_PROVIDER_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
      serviceBudget: constants.SERVICE_PROVIDER_BUDGET,
    }],
    [Arbitrator as AgentClass, {
      tokens: constants.ARBITRATOR_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
      arbitrationCapacity: constants.ARBITRATOR_CAPACITY,
    }],
    [Regulator as AgentClass, {
      tokens: constants.REGULATOR_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
    }],
    [Auditor as AgentClass, {
      tokens: constants.AUDITOR_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
    }],
    [BountyHunter as AgentClass, {
      tokens: constants.BOUNTY_HUNTER_TOKENS,
      reputation: constants.BOUNTY_HUNTER_REPUTATION,
    }],
    [ExternalPartner as AgentClass, {
      tokens: constants.EXTERNAL_PARTNER_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
      votingStrategy: null,
    }],
    [Artist as AgentClass, {
      tokens: constants.ARTIST_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
    }],
    [Collector as AgentClass, {
      tokens: constants.COLLECTOR_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
    }],
    [Speculator as AgentClass, {
      tokens: constants.SPECULATOR_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
    }],
    [StakerAgent as AgentClass, {
      tokens: constants.STAKER_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
    }],
    [PassiveMember as AgentClass, {
      tokens: constants.PASSIVE_MEMBER_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
    }],
    [GovernanceWhale as AgentClass, {
      tokens: constants.GOVERNANCE_WHALE_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
    }],
  ]);

  constructor(simulation: DAOModel) {
    this.simulation = simulation;
  }

  /**
   * Get agents by type using cache for performance
   */
  getAgentsByType(agentType: string): DAOMember[] {
    if (this.cacheDirty || !this.agentTypeCache.has(agentType)) {
      this.rebuildAgentCache();
    }

    return this.agentTypeCache.get(agentType) || [];
  }

  /**
   * Rebuild the agent type cache
   */
  private rebuildAgentCache(): void {
    this.agentTypeCache.clear();

    for (const agent of this.simulation.dao.members) {
      const agentClass = agent.constructor.name;

      if (!this.agentTypeCache.has(agentClass)) {
        this.agentTypeCache.set(agentClass, []);
      }

      this.agentTypeCache.get(agentClass)!.push(agent);
    }

    this.cacheDirty = false;
  }

  /**
   * Mark cache as dirty - call when agents are added/removed
   */
  invalidateCache(): void {
    this.cacheDirty = true;
  }

  /**
   * Create a new agent with appropriate configuration
   */
  createAgent(
    agentClass: AgentClass,
    agentId: string,
    overrides: Partial<AgentConfig> = {}
  ): DAOMember {
    const baseConfig = this.agentConfigs.get(agentClass) || {
      tokens: constants.DEFAULT_AGENT_TOKENS,
      reputation: constants.DEFAULT_AGENT_REPUTATION,
    };

    // Merge base config with overrides
    const config = { ...baseConfig, ...overrides };
    const location = overrides.location || generateRandomLocation();

    if (agentClass === StakerAgent) {
      const totalTokens = Math.max(0, config.tokens || 0);
      const stakedTokens = Math.min(
        totalTokens,
        Math.floor(totalTokens * constants.STAKER_STAKE_RATIO)
      );
      const liquidTokens = Math.max(0, totalTokens - stakedTokens);

      return new StakerAgent(
        agentId,
        this.simulation,
        liquidTokens,
        config.reputation,
        location,
        stakedTokens
      );
    }

    return new agentClass(
      agentId,
      this.simulation,
      config.tokens,
      config.reputation,
      location,
      config.votingStrategy,
      ...Object.values(config).slice(2) // Additional params
    );
  }

  /**
   * Add new members periodically if treasury has funds
   */
  addNewMembers(): void {
    const agentClasses: AgentClass[] = [
      Developer,
      Investor,
      Trader,
      AdaptiveInvestor,
      Delegator,
      LiquidDelegator,
      ProposalCreator,
      Validator,
      ServiceProvider,
      Arbitrator,
      Regulator,
      Auditor,
      BountyHunter,
      ExternalPartner,
      Artist,
      Collector,
      Speculator,
      StakerAgent,
      PassiveMember,
    ];

    for (let i = 0; i < constants.NEW_MEMBERS_PER_INTERVAL; i++) {
      const shouldAdd =
        this.simulation.currentStep > 0 &&
        this.simulation.currentStep % constants.NEW_MEMBER_INTERVAL === 0;

      if (!shouldAdd) {
        continue;
      }

      if (
        this.simulation.dao.treasury.funds < constants.NEW_MEMBER_FUND_REQUIREMENT
      ) {
        continue;
      }

      // Choose random agent type
      const agentClass = randomChoice(agentClasses);

      const newAgent = this.createAgent(
        agentClass,
        `${agentClass.name}_${this.simulation.currentStep}`
      );

      // Add any agent with non-negative reputation (lowered from 25 to enable
      // realistic population dynamics — previously almost all agents were rejected
      // since DEFAULT_AGENT_REPUTATION = 0)
      if (newAgent.reputation >= 0) {
        this.simulation.dao.addMember(newAgent);
        this.simulation.schedule.add(newAgent);
        this.invalidateCache();

        // Deduct cost from treasury
        this.simulation.dao.treasury.withdraw(
          'DAO_TOKEN',
          constants.NEW_MEMBER_COST,
          this.simulation.currentStep
        );
      }
    }
  }

  /**
   * Remove members with negative reputation.
   * NOTE: Creates selection pressure favoring high-initial-reputation agents
   * (GovernanceExpert: 75, RiskManager: 60) over lower-rep types.
   * This is an intentional simplification documented in limitations.
   */
  cullMembers(): void {
    const agentsToRemove: DAOMember[] = [];

    for (const agent of this.simulation.dao.members) {
      if (agent.reputation < 0) {
        agentsToRemove.push(agent);
      }
    }

    for (const agent of agentsToRemove) {
      this.simulation.dao.removeMember(agent);
      this.simulation.schedule.remove(agent);
      this.invalidateCache();
    }
  }

  /**
   * Get agent statistics by type
   */
  getAgentStatistics(): Map<string, number> {
    if (this.cacheDirty) {
      this.rebuildAgentCache();
    }

    const stats = new Map<string, number>();
    this.agentTypeCache.forEach((agents, type) => {
      stats.set(type, agents.length);
    });

    return stats;
  }

  /**
   * Get total agent count
   */
  getTotalAgents(): number {
    return this.simulation.dao.members.length;
  }
}
